### PR TITLE
Gas Condensers

### LINKED ISSUE #11319: Condensing gasses into chemicals should be possible
Though probably not vice-versa without careful balance work.

This means Oxygen, Nitrogen, Plasma (this is somewhat important so maybe make it not condense by normal means), Water Vapor, Nitrous Oxide, Tritium, and CO2 should be convertible to chems via some machinery.

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Resolves #11319

Adds gas condensers, a new machine that converts gases to reagents.
Available round start: should eventually be mapped in chem.

simply pump in a gas and boom reagents. Needs power but otherwise just runs itself.

## Why / Balance

Basically just a nifty little way for chem to get a replenishable supply of basic chems like oxygen, nitrogen, etc.
will be important for le chem rework :trollface: 

## Media

image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added gas condensers. These machines are available at the circuit imprinter and can convert gases into liquids. Tasty!

### CODE REVIEW

**Comment by Aisu9** [comment]
yay, infinite plasma supply. would be fun to have the opposite turning liquid into gas

**Comment by Emisse** [comment]
he doesnt know about the plasma

File: Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                continue;
+
+            var moleToReagentMultiplier = component.MolesToReagentMultiplier;
+            var amount = moles * moleToReagentMultiplier;
+
+            if (_solution.TryAddReagent(uid, solution, gasReagent, amount, out var remaining))
+                continue;
+
+            // if we have leftover reagent, then convert it back to moles and put it back in the mixture.
+            inlet.Air.AdjustMoles(i, remaining.Float() / moleToReagentMultiplier);
```
Comment (Partmedia) [inline Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs:62]: This technically breaks the conservation of energy, i.e. if you do this too often the inlet will mysteriously cool down for no apparent reason. On the other hand, in my personal opinion I don't think this is worth fixing until it proves to be a problem.

File: Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        }
+    }
+
+    public float NumberOfMolesToConvert(ApcPowerReceiverComponent comp, GasMixture mix, float dt)
+    {
+        var hc = _atmosphereSystem.GetHeatCapacity(mix);
+        var alpha = 0.05f; // ratio of heat of vaporization to specific heat, based on water, tune this
```
Comment (Partmedia) [inline Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs:69]:
```suggestion
        var alpha = 0.05f; // tuned to give us 5-ish u/second of reagent conversion
```

File: Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        var energy = comp.Load * dt;
```
Comment (Partmedia) [inline Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs:70]:
```suggestion
        // ignores the energy needed to cool down the solution to the condensation point, but that probably adds too much difficulty and so let's not simulate that
        var energy = comp.Load * dt;
```

File: Resources/Prototypes/Entities/Structures/Piping/Atmospherics/unary.yml [hunk 4]
```diff
@@ -351,3 +351,89 @@
     enabled: true
   - type: ApcPowerReceiver
     powerDisabled: false
+
+- type: entity
+  parent: [ BaseMachinePowered, ConstructibleMachine ]
+  id: BaseGasCondenser
+  name: condenser
+  description: Now we just need some plumbing.
```
Comment (Partmedia) [inline Resources/Prototypes/Entities/Structures/Piping/Atmospherics/unary.yml:359]:
```suggestion
  description: Condenses gases into liquids.
```

**Review by Partmedia — APPROVED** [review]
Wow, this was fast. Conceptual and initial code review proposal from me. I've brought up a potential atmos issue in comments but don't think it's important enough to warrant attention at this time. I'll give it a test in-game soon.

File: Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs (hunk 1 above)
Comment (EmoGarbage404) [reply Content.Server/Atmos/Piping/Unary/EntitySystems/GasCondenserSystem.cs:62]: Yeah i figured its going to be a relatively small amount of moles and I'd rather preserve the mass at the cost of a small bit of heat leaking out.

**Comment by MACMAN2003** [comment]
i think that the condenser could borrow a few mechanics from the gas recycler, namely needing low temperature depending on the gas, like >100c for water vapor to become water

**Comment by EmoGarbage404** [comment]
> i think that the condenser could borrow a few mechanics from the gas recycler, namely needing low temperature depending on the gas, like <100c for water vapor to become water

The intent is that these should be easy enough to operate that they can be set up in the limited space of chem/medical.

Needing a low temperature does impact the rate of condensation, but it would probably be next-to-worthless if it required a more robust setup.

**Comment by Partmedia** [comment]
Comments from my in game testing:

- I tried clicking the device and since nothing happened assumed that it was broken.
- Since I knew you were supposed to use a beaker on this, I clicked my beaker on it and filled it. I kind of expected it to take the beaker in and then require me to eject it, but it literally functions like a water cooler.
- It refills much faster than 5u/sec (remember that atmos updates twice-ish per second), so for all practical purposes this feels like an infinite water cooler. Since the tank is so big *anyway*, I think the condensation rate should be reduced quite a bit.

I don't consider any of these blocking, and especially since this device isn't mapped we can go ahead and merge. If you want to take an opportunity to reconsider the condensation rate, please do.

**Comment by EmoGarbage404** [comment]
rebalanced the machine to approximately 1u/s of condensation off of box distro. Canisters are a tad bit slow but still feels decent. Nerfed the reagents per mol to a bit lower, so now 1 canister == 400u instead of 500.

overall, feels like a nice replenishing backup reserve while not being absurdly strong. I could easily see this getting depleted especially since distro is only ~30% oxygen.

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**